4.4 - ErrorTrap implementation

markstory opened the pull request:
Build up the bulk of the `ErrorTrap` class. This integrations the error renderers, loggers, and callback behavior for what's described in #16228

markstory committed:
Start sketching in the ErrorTrap class

markstory committed:
Add logger & renderer methods.

markstory committed:
Implement the rest of the ErrorTrap behavior.

markstory committed:
Remove redundant import

othercorey commented on src/Error/ErrorRendererInterface.php:
```diff
@@ -31,4 +31,12 @@ interface ErrorRendererInterface
+     * @param string $out The content to output.
+     * @return void
+     */
+    public function output(string $out): void;
```
What do you think about `write` instead? You would write to a stream in most I/O.

markstory committed:
Rename output to write

Also add logging when rendering errors fails.

markstory committed:
Improve typing

markstory committed:
Add more documentation.

markstory committed:
Add test for fallback renderer on construction

othercorey commented on src/Error/ErrorTrap.php:
```diff
@@ -0,0 +1,254 @@
+     *
+     * @return class-string<\Cake\Error\ErrorRendererInterface>
+     */
+    protected function chooseErrorRenderer(): string
```
Why does this return a string instead of the instance?

othercorey replied:
I see this is called in the constructor above. Why don't we do this logic when constructing the renderer instance instead then? 

We have a default, then we choose a default, then we allow setting the renderer anyway and then we instantiate?

Also, kinda seems odd to have setErrorRenderer() when we have a config for it.

markstory replied:
> Also, kinda seems odd to have setErrorRenderer() when we have a config for it.

I don't love the `setConfig()` interface as it has poor typing support and not easy to discover. I would be up to making the methods from `InstanceConfigTrait` into protected ones.

othercorey replied:
So, are we trying to maintain BC support for the `Error` config? If so, having to configure it after instantiation isn't consistent. But, I don't know all the situations you're supporting.

markstory replied:
> So, are we trying to maintain BC support for the Error config?

Yes. The additional configuration methods were for if we wanted to move away from splatting configuration arrays in the future. I could remove those methods for now and add them when we have more concrete use cases for them.

othercorey commented on src/Error/ErrorTrap.php:
```diff
@@ -0,0 +1,254 @@
+        $this->registered = true;
+
+        $level = $this->_config['errorLevel'] ?? -1;
+        error_reporting($level);
```
Was this always coupled with registering? We should document this side-effect if so.

markstory replied:
It has been historically part of the how we bind error handlers.

othercorey commented on src/Error/ErrorTrap.php:
```diff
@@ -0,0 +1,254 @@
+        /** @var \Cake\Error\ErrorRendererInterface $instance */
+        $instance = new $class($this->_config);
+
+        return $instance;
```
Feels a little odd to have this always instantiate. Would you ever need to get the renderer in multiple calls?

markstory replied:
We could store/reuse the renderer as a request could emit multiple renderers. I was trying to avoid more properties and more state but if you think it is important I can switch it up.

othercorey replied:
I vote for simplest design. Just trying to discuss expected convention.

markstory replied:
If the renderers or logger constructor setup logic then I might be concerned with the overhead. Retaining instances means we need attributes, and methods to unset the instances if/when class names change. I think not retaining instances is the simpler design.

ADmad commented on src/Error/ErrorTrap.php:
```diff
@@ -0,0 +1,254 @@
+     * See the `Error` key in you `config/app.php`
+     * for details on the keys and their values.
+     *
+     * @var array
```
```suggestion
     * @var array<string, mixed>
```

ADmad commented on src/Error/ErrorTrap.php:
```diff
@@ -0,0 +1,254 @@
+    /**
+     * Constructor
+     *
+     * @param array $options An options array. See $_defaultConfig.
```
```suggestion
     * @param array<string, mixed> $options An options array. See $_defaultConfig.
```

ADmad commented on src/Error/ErrorTrap.php:
```diff
@@ -0,0 +1,254 @@
+     * @param class-string<\Cake\Error\ErrorLoggerInterface> $class The logging class to use.
+     * @return $this
+     */
+    public function setLogger(string $class)
```
`setLoggerClass()` would be a more intuitive name. At other places in the framework `setLogger()` is used to set an instance.

markstory committed:
Apply suggestions from code review

Co-authored-by: ADmad <ADmad@users.noreply.github.com>

markstory committed:
Rename method.

markstory committed:
Disable output of errors in debug mode for HTML mode

Console will continue to output in debug mode.

markstory committed:
Fix phpcs

markstory committed:
Remove the mutator methods.

We don't have a use case for them yet and they might end up as API
bloat. Fix a few more psalm errors about trace() return value.

markstory committed:
Fix phpcs error

othercorey commented:
Looks ok to me.
